### PR TITLE
nspawn: reduce number of sockets

### PR DESCRIPTION
**Pull request by brauner**
Currently, nspawn opens a multitude of sockets to send various stuff between
the child and parent. This is excessive and wasteful. Just rely on two
well-defined sockets. An outer socket used for synchronization between the
monitor and the outer child and a an inner socket for synchronization between
monitor and inner child.

Note, before this branch can be merged
https://github.com/systemd/systemd/pull/25513
should first be merged.

Signed-off-by: Christian Brauner (Microsoft) <brauner@kernel.org>

### CODE REVIEW

File: src/nspawn/nspawn.c [hunk 1]
```diff
@@ -4898,7 +4871,7 @@
                         if (r < 0)
                                 return log_error_errno(r, "Failed to pick suitable UID/GID range: %m");
 
-                        l = send(uid_shift_socket_pair[0], &arg_uid_shift, sizeof arg_uid_shift, MSG_NOSIGNAL);
+                        l = send(fd_outer_socket_pair[0], &arg_uid_shift, sizeof arg_uid_shift, MSG_NOSIGNAL);
```
[the remark this answers is not in the document]
Comment (brauner) [reply src/nspawn/nspawn.c:4874]: What?

[the remark this answers is not in the document]
Comment (poettering) [reply src/nspawn/nspawn.c:4874]: I think the correct response is "CodeQL, go home, you are drunk!"

File: src/nspawn/nspawn.c [hunk 2]
```diff
@@ -3344,10 +3343,9 @@ static int inner_child(
                 (void) loopback_setup();
 
         if (arg_expose_ports) {
-                r = expose_port_send_rtnl(rtnl_socket);
+                r = expose_port_send_rtnl(fd_inner_socket);
```
Comment (poettering) [inline src/nspawn/nspawn.c:3345]: so, we'll now enqueue the kmsg fifo first on fd_inner_socket, nd the rtnl stuff after that…

File: src/nspawn/nspawn.c [hunk 3]
```diff
@@ -5188,16 +5177,14 @@ static int run_container(
         (void) sd_event_add_signal(event, NULL, SIGCHLD, on_sigchld, PID_TO_PTR(*pid));
 
         if (arg_expose_ports) {
-                r = expose_port_watch_rtnl(event, rtnl_socket_pair[0], on_address_change, expose_args, &rtnl);
+                r = expose_port_watch_rtnl(event, fd_inner_socket_pair[0], on_address_change, expose_args, &rtnl);
```
Comment (poettering) [inline src/nspawn/nspawn.c:5162]: … but here we'll read the rtnl stuff first, and the kmsg stuff second. 

I think the fix is esay: just move the next block that reads fd_kmsg_fifo off the inner socket pair, before this block

File: src/nspawn/nspawn.c [hunk 4]
```diff
@@ -3854,14 +3853,12 @@ static int outer_child(
                 if (r < 0)
                         return r;
 
-                l = send(unified_cgroup_hierarchy_socket, &arg_unified_cgroup_hierarchy, sizeof(arg_unified_cgroup_hierarchy), MSG_NOSIGNAL);
+                l = send(fd_inner_socket, &arg_unified_cgroup_hierarchy, sizeof(arg_unified_cgroup_hierarchy), MSG_NOSIGNAL);
```
Comment (poettering) [inline src/nspawn/nspawn.c]: hmmm, so here we send something on the inner socket, but from the outer process?

File: src/nspawn/nspawn.c [hunk 5]
```diff
@@ -4912,7 +4901,7 @@ static int run_container(
 
         if (arg_unified_cgroup_hierarchy == CGROUP_UNIFIED_UNKNOWN) {
                 /* The child let us know the support cgroup mode it might have read from the image. */
-                l = recv(unified_cgroup_hierarchy_socket_pair[0], &arg_unified_cgroup_hierarchy, sizeof(arg_unified_cgroup_hierarchy), 0);
+                l = recv(fd_inner_socket_pair[0], &arg_unified_cgroup_hierarchy, sizeof(arg_unified_cgroup_hierarchy), 0);
```
Comment (poettering) [inline src/nspawn/nspawn.c]: afaics you can simply change from the inner to the outer fd, to make this systematic. The ordering should already be OK.

**Review by poettering — CHANGES_REQUESTED** [review]
sorry for taking so long. looks vastly simpler. Thanks so much.

File: src/nspawn/nspawn.c (hunk 3 above)
Comment (brauner) [reply src/nspawn/nspawn.c:5162]: done

**Comment by poettering** [comment]
excellent, thanks!

**Pull request merged**